### PR TITLE
Improve canonicalization of generic types with thrown errors

### PR DESCRIPTION
**Pull request by DougGregor**
Noticed by inspection

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/AST/Type.cpp [hunk 1]
```diff
@@ -4168,15 +4168,20 @@ ClangTypeInfo AnyFunctionType::getCanonicalClangTypeInfo() const {
 ASTExtInfo
 AnyFunctionType::getCanonicalExtInfo(bool useClangFunctionType) const {
   assert(hasExtInfo());
+
+  CanGenericSignature genericSig;
+  if (auto *genericFnTy = dyn_cast<GenericFunctionType>(this))
+    genericSig = genericFnTy->getGenericSignature().getCanonicalSignature();
```
Comment (slavapestov) [inline lib/AST/Type.cpp:4174]: getCanonicalSignature() doesn't change the reduced type computation :)

**Review by slavapestov — APPROVED** [review]
Nice catch!

File: lib/AST/Type.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/AST/Type.cpp:4174]: I totally copy-pasta'd this. We can remove it from both here and where I got it from (canonicalization of function types)

**Pull request merged**